### DISCUSSION TITLE
Questions about building graphs with `pos`

### DISCUSSION THREAD

**Discussion by lfangyu09** [pyg-team/pytorch_geometric#7723] [General]
Hi,

Thank you so much for building this library. It really helps me a lot.

May I ask a few questions related to building graphs:

1. When I tried to use the absolute coordinates in `data.pos,` the results were bad. I found in [Issue 3326](https://github.com/pyg-team/pytorch_geometric/issues/3326), you mentioned we could use absolute coordinates in `data.pos` for some GNN layers. Otherwise, you suggested using `edge_attr` to consider the relative distance. Do you have any suggestions about processing the absolute coordinates? 
2. What is the meaning of `data.pos` ? Is it only a key in a dictionary for normal GNN? Only some models like `XConv` could explicitly use `data.pos` ?
3. For the node feature, I have a feature that is continuous. It does not change across different nodes but changes across different time steps. Do you have any suggestions about processing this feature?
4. For the edge feature, I have a feature that is similar to the node type. It does not change across different time steps but does change across different nodes. Do you have any suggestions about processing this feature? If I want to set this feature as the edge type, how can I do it?

Thank you so much for your reply! I really appreciate your help!

**Comment by rusty1s** [comment]
1. Both ways can be found in literature. In general, using absolute coordinates no longer makes your GNN layer translation invariant, and thus might not generalize well to positions never seen before during training. If positions are equally distributed over training and test sets, using absolute coordinates should work quite well as well though.
2. Yes, it is just a separate attribute to differentiate node features from node positions. Following (1), you can also do `x = torch.cat([data.x, data.pos], dim=-1)` and interpret them as node features.
3. You should be able to just use it as a regular feature.
4. You can encode it as a one-hot vector if it models a type/index. Alternatively, there exist GNN layers that can take `edge_type` information into account, e.g., `RGCNConv`.

**Reply by lfangyu09** [reply]
Thank you so much for your helpful reply. 

I have another question about encoding type/index. You suggest using one-hot to encode. I saw some papers that used embedding (e.g., `torch.nn.Embedding`) to encode. Which one is better for encoding type/index? Thanks.

**Reply by rusty1s** [reply]
Usually, the `Embedding` approach is better since it is more memory and computation friendly. You avoid to store the one-hot matrix and avoid computing the embeddings via matrix multiplication.

**Reply by lfangyu09** [reply]
Thank you so much for your helpful reply.

I have a question about `data.pos`. I want to do the node regression, while I want to explicitly use `data.pos`. I checked some models, such as `PointTransformerConv`, but they did not include `edge_attr`. May I ask which model could explicitly consider `data.x`, `data.pos`, and `data.edge_ettr`? Thanks.

**Reply by rusty1s** [reply]
How do you want to use `data.pos`? I suggest to either include it into `x` (as a feature to encode absolute positions) or `edge_attr` (as a feature to encode relative positions between nodes).

**Reply by lfangyu09** [reply]
I used to use `GNS` model from the paper (_Learning to Simulate Complex Physics with Graph Networks_). It did (1) `data.x= torch.cat((data.x, data.pos), dim=-1)` and (2) use relative positions between nodes in `edge_attr`. 
I would like to try other models, such as `PointTransformerConv`. However, these models actually include `data.pos` into `x` or `edge_attr`. Is it correct? Thanks.

**Reply by rusty1s** [reply]
Yes, this will happen internally.
